library/std/src/thread/local.rs

```diff
@@ -267,6 +267,8 @@ macro_rules! __thread_local_inner {
                         if let $crate::option::Option::Some(init) = _init {
                             if let $crate::option::Option::Some(value) = init.take() {
                                 return value;
+                            } else if $crate::cfg!(debug_assertions) {
+                                unreachable!("missing initial value");
```

Comment:
Note that I was suggesting the `debug_unreachable!` to get the `unsafe hint::unreachable_unchecked` optimization to elide the extra branch. May not be worth for a first implementation, to be sure (I'll let T-libs{-api} be the judge of that), but I prefer pointing it out anyways 🙂

Reply:
I suspect eliding that one check is completely insignificant compared to everything else that's going on around thread local initialization. I don't think it's worth the extra unsafety.